### PR TITLE
[libc] Only declare float128 math functions in the generated math.h if float128 type is supported.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/include/math.h.def [hunk 1]
```diff
@@ -11,6 +11,26 @@
 
 #include <__llvm-libc-common.h>
 #include <llvm-libc-macros/math-macros.h>
+#include <llvm-libc-types/float128.h>
+
+#ifdef LIBC_COMPILER_HAS_FLOAT128
```
Comment (SchrodingerZhu) [inline libc/include/math.h.def:16]: is the macro exposed to user interface?

Comment (lntue) [reply libc/include/math.h.def:16]: Yes they are.  All the headers inside `libc/include` are expected to be used by users in full build mode.

File: libc/include/math.h.def [hunk 2]
```diff
@@ -11,6 +11,26 @@
...
+// Only declare float128 functions if float128 is supported.
+
+__BEGIN_C_DECLS
+
+float128 copysignf128(float128) __NOEXCEPT;
```
Comment (michaelrj-google) [inline libc/include/math.h.def:21]: With the generated headers we shouldn't be defining function prototypes except through the generation. Is there a way we could do this with exclusions/inclusions for the entrypoints list? We could do something similar to `has_sys_random`.

Comment (lntue) [reply libc/include/math.h.def]: I updated the patch to add guards on generated public api instead.

**Comment by michaelrj-google** [comment]
I don't understand why we need to have these guards at compile time and we can't check if the compiler supports float128 when we're generating the headers. Was there a patch I missed with discussion that explained that?

**Comment by lntue** [comment]
> I don't understand why we need to have these guards at compile time and we can't check if the compiler supports float128 when we're generating the headers. Was there a patch I missed with discussion that explained that?

This came up during the discussion of https://github.com/llvm/llvm-project/pull/80682, namely, the distinction between compilers used to build our libc vs. the compilers of the users of the generated libc / headers.  Without the guards for these functions, if the compiler used to build the libc has `float128` support, it will force users of generated headers to use compilers supporting `float128` also.  Otherwise, it will error out whenever the generated `math.h` is included transitively.

File: libc/utils/HdrGen/PublicAPICommand.cpp [hunk 3]
```diff
@@ -102,6 +102,12 @@ void writeAPIFromIndex(APIIndexer &G,
     llvm::Record *RetValSpec = FunctionSpec->getValueAsDef("Return");
     llvm::Record *ReturnType = RetValSpec->getValueAsDef("ReturnType");
 
+    bool Guarded =
```
Comment (michaelrj-google) [inline libc/utils/HdrGen/PublicAPICommand.cpp:105]: nit: Add a comment that ideally we would group functions based on their guarding macro.

**Review by michaelrj-google — APPROVED** [review]
Based on your explanation I think you're right that this is the correct solution. Given that we are doing this, we should probably provide a flag to allow users to manually enable float128 instead of just basing it on if the current compiler supports it.

File: libc/utils/HdrGen/PublicAPICommand.cpp (hunk 3 above)
Comment (lntue) [reply libc/utils/HdrGen/PublicAPICommand.cpp:107]: Done.

**Pull request merged**